### ISSUE TITLE
The `router` module is not an IBCMiddleware

### ISSUE THREAD

**Issue by damiannolan** [strangelove-ventures/packet-forward-middleware#55]
## Problem Description

Context: This is something that arose when I was asked to review and help out the gaia team with debugging an upgrade handler. I reviewed a lot of the `app` configuration code in the gaia repo and came to find that the transfer keeper is misconfigured to bypass the `router` module on `SendPacket`. 

For an IBC application stack composed of `transfer` and `router` there is two distinct flows. For the sake of simplicity let's reduce this to inbound and outbound packet flows. i.e. `RecvPacket` and `SendPacket`.

1. When receiving a packet the expected flow should be: ibc-core -> router -> transfer.
2. When sending a packet the expected flow should be: transfer -> router -> ibc-core.

In gaia number 1 holds true due to the [wrapping of the transfer module and correctly binding it using the ibc router](https://github.com/cosmos/gaia/blob/main/app/keepers/keepers.go#L323-L329).

However, number 2 does not respect the expected flow above, as the [transfer keeper is misconfigured for the `ICS4Wrapper` argument here](https://github.com/cosmos/gaia/blob/main/app/keepers/keepers.go#L284). When a user sends an ibc `MsgTransfer`, it will call into [`SendPacket` on the `ICS4Wrapper`](https://github.com/cosmos/ibc-go/blob/b8e05d39a66ad5eebc01f1c2c8627d524ab9cd4d/modules/apps/transfer/keeper/relay.go#L136), and with its current configuration it bypasses the `router` module completely, and is calling directly into ibc core's channel keeper.

At the moment this is technically fine, as the `router` module doesn't implement any logic on `SendPacket`. However, the issue becomes apparent when trying to compose an application with more middlewares (e.g. `29-fee`). Consider an application stack of transfer -> router -> fee. The fee middleware module does implement logic on `SendPacket` so bypassing it would cause some very strange issues.

Currently there is only [a type assertion for `router` to implement the `IBCModule` interface](https://github.com/strangelove-ventures/packet-forward-middleware/blob/main/router/module.go#L36). 

## Proposal

The `router` module should implement the [`IBCMiddleware`](https://github.com/cosmos/ibc-go/blob/b8e05d39a6/modules/core/05-port/types/module.go#L137) interface and allow the `RouterKeeper` to be passed to `NewTransferKeeper` as the `ICS4Wrapper` argument. 

This will allow ibc application stacks including the `router` module to be configured correctly, and adhere to the conventional flows.

## Outcomes

The ibc transfer keeper can be configured like so:

```diff
appKeepers.TransferKeeper = ibctransferkeeper.NewKeeper(
	appCodec,
	appKeepers.keys[ibctransfertypes.StoreKey],
	appKeepers.GetSubspace(ibctransfertypes.ModuleName),
-	appKeepers.IBCKeeper.ChannelKeeper,
+	appKeepers.RouterKeeper,
	appKeepers.IBCKeeper.ChannelKeeper,
	&appKeepers.IBCKeeper.PortKeeper,
	appKeepers.AccountKeeper,
	appKeepers.BankKeeper,
	appKeepers.ScopedTransferKeeper,
)
```

**Comment by agouin** [comment]
The packet-forward-middleware already implements the `ICS4Wrapper` in `v3` and `v4` branches. One of our next tasks is to cherry-pick this forward into `v5` and `main` which targets ibc-go v6. 

We will have releases shortly, starting with `v3` and `v4` to include this and other latest changes.

Here's an example of the middleware implementation in addition to ics-29 fee middleware:
https://github.com/strangelove-ventures/gaia/tree/andrew/pfm_v9

**Comment by damiannolan** [comment]
Cool! Feel free to close my PR to main then :)

**Comment by agouin** [comment]
Thank you for the PR! But yes I think we will close it and cherry-pick forward to keep things consistent across branches.

**Comment by agouin** [comment]
Resolved in v3.1.0+, v4.0.0+, v5.1.0+, and v6.0.0+
